### DISCUSSION TITLE
Mouse Tool dispatchAction change with latest glsp-client update

### DISCUSSION THREAD

**Discussion by ivy-lli** [eclipse-glsp/glsp#616] [General] [answered]
Is there a reason why the `MouseTool#dispatchAction` change was made in the sprotty update refactoring (https://github.com/eclipse-glsp/glsp-client/commit/88e4725bd27f2ebf0fcd925c3e9be16437725159#diff-cf89edacf90dd71f9f8c27174ec5b956393b6b3e40f83c6eaf63649c53d963f1L85-R86)?

The reason if this question is that I tried to update our client also to the latest glsp-client and now I have some strange mouse feedback.

My problem is if I try to create a new edge and hover the source, the evaluation is started in the EdgeCreationTool if the hovered element is a valid soure and if so, a coursor feedback action is created. 
But with the same hover also other actions are created, e.g. the popup action (which is a Promise Action, because it waits for a specific time before the popup should really shown up). 
All ok for me. But now the MouseTool collects all this actions (in the dispatchAction method) and dispatch them over the actionDispatcher. And if the actions now collects a Promise<Action> and then the CursorFeedbackAction, it waits now first for the Promise until the other action is dispatched right? 
Or do I understand this change wrong? Because for my understanding, in the old code it waited only for the Promise<Action>'s and the others it dispatched immediately. Or did I something wrong with the update?

**Comment by tortmayr** [comment, marked as answer]
You are right this looks like an over optimization to me. I the old code was only blocking in the synchronous use case. i.e. if the object to dispatch is an action we directly dispatched it and waited for the result. If the object to dispatch was a Promise of an action we did no wait for the dispatching result.
With the new code we always  wait until the `actionDispatcher.dispatch()` method resolves. This could definitely cause the issues you described.
Even though we didn't encounter an issues in the workflow examples it might still be better to revert this change.
I'm currently also preparing an update of the `@eclipse-glsp/ide` module which should be ready soon.
So If you can postpone your update a little bit you will be able to consume a "fixed" version of the GLSP  client and an updated ide package. This should make updating you code base relatively straight forward.

**Reply by ivy-lli** [reply]
Cool Thanks :+1:
